api/dataset.go

```diff
@@ -1070,6 +1076,73 @@ func (d *PublishCheck) Check(handle func(http.ResponseWriter, *http.Request)) ht
 
 		if versionDoc != nil {
 			if versionDoc.State == models.PublishedState {
+				defer func() {
+					if err := r.Body.Close(); err != nil {
+						log.ErrorC("could not close response body", err, nil)
+					}
+				}()
+
+				currentVersion, err := models.CreateVersion(r.Body)
+				if err != nil {
+					log.ErrorC("failed to model version resource based on request", err, log.Data{"dataset_id": id, "edition": edition, "version": version})
+					http.Error(w, err.Error(), http.StatusBadRequest)
+					return
+				}
+
+				// We can allow public download links to be modified by the exporters when a version is published.
+				// Note that a new version will be created which contain only the download information to prevent
+				// any forbidden fields from being set on the published version
+				if currentVersion.Downloads != nil {
+					if currentVersion.Downloads.CSV != nil && currentVersion.Downloads.CSV.Public != "" {
+						newVersion := &models.Version{
+							Downloads: &models.DownloadList{
+								CSV: &models.DownloadObject{
+									Public: currentVersion.Downloads.CSV.Public,
+									Size:   currentVersion.Downloads.CSV.Size,
+									URL:    currentVersion.Downloads.CSV.URL,
+								},
+							},
+						}
+
+						b, err := json.Marshal(newVersion)
```

Comment:
Would it not be better to build up the versionDoc before marshalling; this will reduce the duplicate code and also allow for requests that want to update both csv and xls public downloads at the same time? I know the web does not do this and API customers can't do this as it is a private endpoint but why limit what our applications can do in the future, for example alerts which will need to be added in the future may change with a new download links, who knows?